### DISCUSSION TITLE
How do I retain actual object that is passed from a resolver to a type?

### DISCUSSION THREAD

**Discussion by djpate** [MichalLytek/type-graphql#1747] [Q&A]
Assume I have the following resolver
```ts
@Resolver(Post)
export class PostResolver {

  @Query(returns => [Post])
  async posts(@Ctx() context: KaidaRequestContext) {
    const models = await Post.where({where: {userId: context.user.member_id}})
    console.log(models[0].id) #=> return correctly
    return models
  }
}
```

and I have a post type as such
```ts
@ObjectType()
export class Post {
  @Field(type => ID)
  id: string;
}
```

This works perfectly fine.

Now what if I try to create a custom method for my id field such as 

```ts
@ObjectType()
export class PhishingTest {
  @Field(type => ID)
  id(): sting => {
    return `${this.id}-foobar`
  }
```

now `this.id` is `undefined`.

The reason is that my `Post` model where I do my data fetching is a class that looks like this

```ts
Class Post
   get id() {
     return this.id
   }

  constructor(protected data) {
  }
}
```

So I'm assuming somewhere in the context setting of the type the this gets mumbled up? Not sure what do to about that.

**Comment by carlocorradini** [comment]
@djpate I don't understand what you want to achieve.
Can you add more information because there is only the `id` field/property and the example is not clear.
Do you want to add a field to an ObjectType from another resolve?

**Comment by djpate** [comment]
I'm trying to implement something similar to the averageRating method in the examples.

```ts
@Field(type => Float, { nullable: true })
  averageRating(@Arg("since") sinceDate: Date): number | null {
    const ratings = this.ratings.filter(rate => rate.date > sinceDate);
    if (!ratings.length) return null;

    const ratingsSum = ratings.reduce((a, b) => a + b, 0);
    return ratingsSum / ratings.length;
  }
```

in the method above, `this.ratings` is availble. but for my code, `this` is not the instance of the actual object that is being resolved.

so

```ts
@ObjectType()
export class PhishingTest {
  @Field(type => ID)
  id(): sting => {
    return `${this.id}-foobar` <---- this.id here is undefined 
  }
```

**Reply by carlocorradini** [reply]
In the ObjectType, where is defined the `id` property? I don't see any
For example, if ObjectType has a property called `_id: number` (private and not Field) and a method (public and Field) it should work fine if `_id` is always present

**Reply by djpate** [reply]
It's defined on my Post Class.

```ts
Class Post
   get id() {
     return this.id
   }

  constructor(protected data) {
  }
}```

my resolver returns an array of Post so I'm expecting that in my object type, I can accept the actual object that's passed in

**Comment by djpate** [comment]
To be clear, this works fine If I use a FieldResolver() on the resolver itself. I'd much rather use it on the type thought

**Comment by djpate** [comment]
OK i've figured it out. I need to use @Root() as an argument to my field method in the ObectType and that gives me access to the actual object.
